mypy checks are failing with:

```
[2](https://github.com/pypa/setuptools/actions/runs/28333781829/job/83936383357?pr=5244#step:11:63)
[gw0] darwin -- Python 3.14.6 /Users/runner/work/setuptools/setuptools/.tox/py/bin/python
setuptools/tests/test_config_discovery.py:24: error: Need type annotation for "greedy_product"  [var-annotated]
setuptools/tests/test_config_discovery.py:24: error: Argument 1 to "compose" has incompatible type "type[list[_T]]"; expected "Callable[[product[tuple[Never]]], list[Never]]"  [arg-type]
setuptools/tests/test_config_discovery.py:176: error: Too many arguments  [call-arg]
```

--- a/setuptools/tests/test_config_discovery.py
+++ b/setuptools/tests/test_config_discovery.py
@@ -1,12 +1,12 @@
 import os
 import sys
+from collections.abc import Iterable
 from configparser import ConfigParser
 from itertools import product
 from typing import cast
 
 import jaraco.path
 import pytest
-from jaraco.functools import compose
 from path import Path
 
 import setuptools  # noqa: F401 # force distutils.core to be patched
@@ -21,15 +21,17 @@ from .textwrap import DALS
 
 import distutils.core
 
-greedy_product = compose(list, product)
-"""
-``itertools.product`` rendered eager (a list rather than a one-shot iterator).
 
-``@pytest.mark.parametrize`` needs a re-iterable Collection for ``argvalues``:
-a bare iterator gets exhausted after the first of a class' test methods is
-collected, silently skipping the rest. Deprecated in pytest 9.1, raising
-``PytestRemovedIn10Warning`` (pytest-dev/pytest#13409).
-"""
+def greedy_product(*iterables: Iterable[object]) -> list[tuple[object, ...]]:
+    """
+    ``itertools.product`` rendered eager (a list rather than a one-shot iterator).
+
+    ``@pytest.mark.parametrize`` needs a re-iterable Collection for ``argvalues``:
+    a bare iterator gets exhausted after the first of a class' test methods is
+    collected, silently skipping the rest. Deprecated in pytest 9.1, raising
+    ``PytestRemovedIn10Warning`` (pytest-dev/pytest#13409).
+    """
+    return list(product(*iterables))
 
 
 class TestFindParentPackage:

--- a/setuptools/tests/test_config_discovery.py
+++ b/setuptools/tests/test_config_discovery.py
@@ -1,12 +1,12 @@
 import os
 import sys
-from collections.abc import Iterable
 from configparser import ConfigParser
 from itertools import product
 from typing import cast
 
 import jaraco.path
 import pytest
+from jaraco.functools import compose
 from path import Path
 
 import setuptools  # noqa: F401 # force distutils.core to be patched
@@ -21,17 +21,19 @@ from .textwrap import DALS
 
 import distutils.core
 
+greedy_product = compose(list, product)  # type: ignore[var-annotated,arg-type] # python/mypy#13540
+"""
+``itertools.product`` rendered eager (a list rather than a one-shot iterator).
 
-def greedy_product(*iterables: Iterable[object]) -> list[tuple[object, ...]]:
-    """
-    ``itertools.product`` rendered eager (a list rather than a one-shot iterator).
+``@pytest.mark.parametrize`` needs a re-iterable Collection for ``argvalues``:
+a bare iterator gets exhausted after the first of a class' test methods is
+collected, silently skipping the rest. Deprecated in pytest 9.1, raising
+``PytestRemovedIn10Warning`` (pytest-dev/pytest#13409).
 
-    ``@pytest.mark.parametrize`` needs a re-iterable Collection for ``argvalues``:
-    a bare iterator gets exhausted after the first of a class' test methods is
-    collected, silently skipping the rest. Deprecated in pytest 9.1, raising
-    ``PytestRemovedIn10Warning`` (pytest-dev/pytest#13409).
-    """
-    return list(product(*iterables))
+The ``type: ignore`` comments work around python/mypy#13540: mypy collapses the
+overloaded ``product`` to its first overload when it is passed by value through
+``compose``'s ParamSpec, mis-typing the result as a single-argument callable.
+"""
 
 
 class TestFindParentPackage:
@@ -175,7 +177,7 @@ class TestDiscoverPackagesAndPyModules:
 
     @pytest.mark.parametrize(
         ("config_file", "param", "circumstance"),
-        greedy_product(
+        greedy_product(  # type: ignore[call-arg] # python/mypy#13540
             ["setup.cfg", "setup.py", "pyproject.toml"],
             ["packages", "py_modules"],
             FILES.keys(),

